lightningd/bitcoind.c

```diff
@@ -420,59 +119,74 @@ struct estimatefee {
+				const jsmntok_t *idtok,
+				struct estimatefee_call *call)
 {
+	const jsmntok_t *resulttok, *feeratetok;
 	u64 feerate;
-	struct estimatefee *efee = bcli->cb_arg;
+
+	resulttok = json_get_member(buf, toks, "result");
+	if (!resulttok)
+		bitcoin_plugin_error(call->bitcoind, buf, toks,
+				     "getfeerate",
+				     "bad 'result' field");
+
+	feeratetok = json_get_member(buf, resulttok, "feerate");
+	if (!feeratetok)
+		bitcoin_plugin_error(call->bitcoind, buf, toks,
+				     "getfeerate",
+				     "bad 'feerate' field");
 
 	/* FIXME: We could trawl recent blocks for median fee... */
-	if (!extract_feerate(bcli, bcli->output, bcli->output_bytes, &feerate)) {
-		log_unusual(bcli->bitcoind->log, "Unable to estimate %s/%u fee",
-			    efee->estmode[efee->i], efee->blocks[efee->i]);
+	if (!json_to_u64(buf, feeratetok, &feerate)) {
+		log_unusual(call->bitcoind->log, "Unable to estimate %s/%u fee",
+			    call->estmode[call->i], call->blocks[call->i]);
 
 #if DEVELOPER
 		/* This is needed to test for failed feerate estimates
 		 * in DEVELOPER mode */
-		efee->satoshi_per_kw[efee->i] = 0;
+		call->satoshi_per_kw[call->i] = 0;
 #else
 		/* If we are in testnet mode we want to allow payments
 		 * with the minimal fee even if the estimate didn't
 		 * work out. This is less disruptive than erring out
 		 * all the time. */
 		if (chainparams->testnet)
-			efee->satoshi_per_kw[efee->i] = FEERATE_FLOOR;
+			call->satoshi_per_kw[call->i] = FEERATE_FLOOR;
 		else
-			efee->satoshi_per_kw[efee->i] = 0;
+			call->satoshi_per_kw[call->i] = 0;
 #endif
 	} else
 		/* Rate in satoshi per kw. */
-		efee->satoshi_per_kw[efee->i]
+		call->satoshi_per_kw[call->i]
 			= feerate_from_style(feerate, FEERATE_PER_KBYTE);
 
-	efee->i++;
-	if (efee->i == tal_count(efee->satoshi_per_kw)) {
-		efee->cb(bcli->bitcoind, efee->satoshi_per_kw, efee->arg);
-		tal_free(efee);
+	call->i++;
+	if (call->i == tal_count(call->satoshi_per_kw)) {
+		call->cb(call->bitcoind, call->satoshi_per_kw, call->arg);
+		tal_free(call);
 	} else {
 		/* Next */
-		do_one_estimatefee(bcli->bitcoind, efee);
+		do_one_estimatefee(call->bitcoind, call);
 	}
-	return true;
 }
 
 static void do_one_estimatefee(struct bitcoind *bitcoind,
-			       struct estimatefee *efee)
+			       struct estimatefee_call *call)
 {
-	char blockstr[STR_MAX_CHARS(u32)];
-
-	snprintf(blockstr, sizeof(blockstr), "%u", efee->blocks[efee->i]);
-	start_bitcoin_cli(bitcoind, NULL, process_estimatefee, false,
-			  BITCOIND_LOW_PRIO,
-			  NULL, efee,
-			  "estimatesmartfee", blockstr, efee->estmode[efee->i],
-			  NULL);
+	struct jsonrpc_request *req;
+
+	req = jsonrpc_request_start(bitcoind, "getfeerate",
+				    bitcoind->log, getfeerate_callback,
+				    call);
+	json_add_num(req->stream, "blocks", call->blocks[call->i]);
+	json_add_string(req->stream, "mode", call->estmode[call->i]);
+	jsonrpc_request_end(req);
+	plugin_request_send(strmap_get(&bitcoind->pluginsmap,
+				       "getfeerate"), req);
```

Comment:
Would be nice if we could bundle these and send all the block/mode combinations to the plugin at once. That'd allow us to have a single all-or-nothing handler and not having to loop in `lightningd`.

Reply:
Indeed. As a cleanup or before the release ?

Reply:
Post.  Too late for release now.

Reply:
Opened https://github.com/ElementsProject/lightning/issues/3508 to track this.